### DISCUSSION TITLE
model_validator to return None

### DISCUSSION THREAD

**Discussion by CharlesPerrotMinot** [pydantic/pydantic#7215] [Question] [answered]
Hi!

The following behavior used to work (with previous v2 versions of Pydantic, see https://github.com/pydantic/pydantic/issues/6474) but now it doesn't any more.
My goal is to have a validator which checks that all of the attributes are not None, and if one is, return None for the whole object. To do so, I'm using a `model_validator`, but it fails at asserting that it's None, and trying to print the resulting object triggers an error related to `__pydantic_extra__`

```
from typing import Any, Optional
from pydantic import (
    BaseModel,
    model_validator,
)

class Position(BaseModel):
  latitude: float
  longitude: float

  @model_validator(mode="wrap")
  @classmethod
  def validate_all_values(
      cls,
      kwargs: dict[str, Any],
      handler,
  ) -> Optional["Position"]:
      if all(kwargs.values()):
          return handler(kwargs)
      return None

position = Position(**
  {
    "latitude": 40.2,
    "longitude": None,
  }
)
print(position)
assert position is None
```

If I try to execute it, I get:
```
Traceback (most recent call last):
  File "test.py", line 40, in <module>
    print(position)
  File "/opt/homebrew/lib/python3.11/site-packages/pydantic/main.py", line 885, in __str__
    return self.__repr_str__(' ')
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pydantic/_internal/_repr.py", line 55, in __repr_str__
    return join_str.join(repr(v) if a is None else f'{a}={v!r}' for a, v in self.__repr_args__())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pydantic/_internal/_repr.py", line 55, in <genexpr>
    return join_str.join(repr(v) if a is None else f'{a}={v!r}' for a, v in self.__repr_args__())
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pydantic/main.py", line 873, in __repr_args__
    pydantic_extra = self.__pydantic_extra__
                     ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/pydantic/main.py", line 715, in __getattr__
    pydantic_extra = object.__getattribute__(self, '__pydantic_extra__')
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Position' object has no attribute '__pydantic_extra__'. Did you mean: '__pydantic_private__'?
```

**Comment by harunyasar** [comment]
As far as I understand, you shouldn't return `None` type from a function decorated with `@model_validator` it is supposed to either return a model or raise an error.

```python
from typing import Any, Optional

from pydantic import BaseModel, model_validator


class Position(BaseModel):
    latitude: float
    longitude: Optional[float]

    @model_validator(mode="wrap")
    @classmethod
    def validate_all_values(cls, values, handler):
        if all(values.values()):
            return handler(values)
        raise ValueError("Something raised")


position = Position(
    **{
        "latitude": 40.2,
        "longitude": None,
    }
)
```

If you want to return `None` type after a validation, the best option would be to catch the error and return a `None` type.

```python
try:
    position = Position(
    **{
        "latitude": 40.2,
        "longitude": None,
    }
)
except ValidationError:
    return None
```

**Comment by CharlesPerrotMinot** [comment]
Thank you @harunyasar!
I'm a bit surprised since @samuelcolvin himself recommended to return None from the validator: https://github.com/pydantic/pydantic/issues/6474#issuecomment-1623072370

Also, having to catch the None every time the Model is instantiated is quite bothersome and not ideal, and I would rather being able to not have to put the attributes as optional, since they should never be.

**Reply by harunyasar** [reply]
So, you found the solution then, didn't you? You could replace the `ValueError` line with `None`. It will work this way too.

When I was writing this reply I have noticed what you meant. Will have a look at it again.

**Reply by CharlesPerrotMinot** [reply]
Sadly no. It used to work, but doesn't any more, and raises the error in the first comment when I try to print it, knowing it's clearly not a None object: `AttributeError: 'Position' object has no attribute '__pydantic_extra__'. Did you mean: '__pydantic_private__'?`

**Reply by harunyasar** [reply]
I was trying to find the issue with `__pydantic_extra__` but I have not been able to find any clue. There is one possible solution in my mind that is overriding `__new__` method. It could be something as follows:

```python
class Position(BaseModel):
    latitude: float
    longitude: float

    def __new__(cls, *args, **kwargs):
        if not all(kwargs.values()):
            return None
        return super().__new__(cls)


position = Position(
    **{
        "latitude": 40.2,
        "longitude": None,
    }
)
```

If you ask me I wouldn't override `__new__` method because there might be unforeseen consequences. I hope that @samuelcolvin jump into the issue and point us the optimal solution.

**Reply by CharlesPerrotMinot** [reply]
It won't work, because `__new__` isn't called any more with nested models. See https://github.com/pydantic/pydantic/issues/6474#issuecomment-1623072370

**Comment by CharlesPerrotMinot** [comment, marked as answer]
Pretty sure #6820 is solving the issue, in particular https://github.com/pydantic/pydantic/pull/6820/commits/4587cfd66c9166bb0a603350011f0fffdf7f487d
Will test as soon as 2.4 is out!
